### DISCUSSION TITLE
Add folder vs Add Data for Flask Application

### DISCUSSION THREAD

**Discussion by aadejare** [pyinstaller/pyinstaller#8353] [Help] [answered]
Hello,

I wanted to know when to use certain features for the flask app I am trying to convert into a desktop app.  
The documentation, at least to me, is not clear on how to think about what to use. Normally, I use add folder for the static and the template folder. Do I have to go and add items within the folder that are not python scripts (images, html files, doc files)?  What if the application will create new folders and upload images which will need to be access?

**Comment by bwoodsend** [comment]
Surely there shouldn't be Python scripts in static/templates directories?

**Comment by aadejare** [comment]
There are no python scripts there, but there are static files and template files that you have to worry about

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Reply by rokm** [reply, marked as answer]
You need to ensure that all resources that your application uses are collected; and you usually need to ensure that yourself (since PyInstaller's dependency tracking is primarily about python module import tree, and binary dependencies of python extensions). 

If you use `--add-data` to collect a directory, then its contents are collected recursively and you do not need to worry about individual items in there. Is this what you were asking?

> What if the application will create new folders and upload images which will need to be access?

Then the application will crash miserably if it is installed on read-only filesystem or with read-only permissions. Or if you happen to be making macOS .app bundle, which is supposed to be immutable. Or, if you are using onefile mode, the uploaded files will be lost between application restarts, since application data directory is ephemeral.

**Reply by aadejare** [reply]
If I am understanding it correctly,
If there is a folder called static and a subfolder called images
If I add the static directory it will get all of the FILES within static?
What about the sub directory, will it get all of the files within images?

**Reply by rokm** [reply]
Yes, if you use `--add-data path/to/static:.` (or its equivalent .spec file directive), the `static` directory should be collected recursively - all files and all subdirectories in it, and all files and subdirectories in those subdirectories, and so on...

**Reply by aadejare** [reply]
Thank you Rokm, I appreciate it and it was helpful.  Finally, about the Mac question, can you do one folder mode for the Mac and address the issue you talked about?

**Reply by rokm** [reply]
Sure, you can make a "regular" onedir build instead of an .app bundle on Mac, too. But I guess that might depend on what you plan to do with your application (i.e., who your users are and how you plan to distribute it).
